crates/flipperzero/src/furi/thread.rs

```diff
@@ -13,3 +136,132 @@ pub fn sleep(duration: core::time::Duration) {
         }
     }
 }
+
+/// A unique identifier for a running thread.
+#[cfg(feature = "alloc")]
+pub struct ThreadId(sys::FuriThreadId);
+
+/// A handle to a thread.
+#[cfg(feature = "alloc")]
+pub struct Thread {
+    /// Guaranteed to be UTF-8.
+    name: Option<CString>,
+    thread: *mut sys::FuriThread,
+}
+
+#[cfg(feature = "alloc")]
+impl Drop for Thread {
+    fn drop(&mut self) {
+        // TODO: is this safe to do while the thread is running?
+        unsafe { sys::furi_thread_free(self.thread) };
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl Thread {
+    fn new(name: Option<CString>, stack_size: Option<usize>) -> Self {
+        let stack_size = stack_size.unwrap_or(MIN_STACK_SIZE);
+
+        unsafe {
+            let thread = sys::furi_thread_alloc();
+            if let Some(name) = name.as_deref() {
+                sys::furi_thread_set_name(thread, name.as_ptr());
+            }
+            sys::furi_thread_set_stack_size(thread, stack_size);
+            Thread { name, thread }
+        }
+    }
+
+    /// Gets the thread's unique identifier.
+    ///
+    /// Returns `None` if the thread has terminated.
+    pub fn id(&self) -> Option<ThreadId> {
+        // TODO: The Rust stdlib generates its own unique IDs for threads that are valid
+        // even after a thread terminates.
+        let id = unsafe { sys::furi_thread_get_id(self.thread) };
+        if id.is_null() {
+            None
+        } else {
+            Some(ThreadId(id))
+        }
+    }
+
+    /// Gets the thread's name.
+    ///
+    /// Returns `None` if the thread has terminated, or is unnamed, or has a name that is
+    /// not valid UTF-8.
+    pub fn name(&self) -> Option<&str> {
+        self.cname()
+            .map(|s| unsafe { str::from_utf8_unchecked(s.to_bytes()) })
+    }
+
+    fn cname(&self) -> Option<&CStr> {
+        self.name.as_deref()
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl fmt::Debug for Thread {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Thread")
+            .field("name", &self.name())
+            .finish_non_exhaustive()
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl ufmt::uDebug for Thread {
+    fn fmt<W>(&self, f: &mut ufmt::Formatter<'_, W>) -> Result<(), W::Error>
+    where
+        W: ufmt::uWrite + ?Sized,
+    {
+        // TODO: ufmt doesn't provide an impl of uDebug for &str.
```

Comment:
That seems quite the omission. 😦

Reply:
The upstream issue is https://github.com/japaric/ufmt/issues/52.